optuna/multi_objective/trial.py

```diff
@@ -0,0 +1,357 @@
+        """
+
+        return self._trial.distributions
+
+    @property
+    def user_attrs(self) -> Dict[str, Any]:
+        """Return user attributes.
+
+        Returns:
+            A dictionary containing all user attributes.
+        """
+
+        return self._trial.user_attrs
+
+    @property
+    def system_attrs(self) -> Dict[str, Any]:
+        """Return system attributes.
+
+        Returns:
+            A dictionary containing all system attributes.
+        """
+
+        return self._trial.system_attrs
+
+    @property
+    def datetime_start(self) -> Optional[datetime]:
+        """Return start datetime.
+
+        Returns:
+            Datetime where the :class:`~optuna.trial.Trial` started.
+        """
+
+        return self._trial.datetime_start
+
+    # TODO(ohta): Add `to_single_objective` method.
+    # This method would be helpful to use the existing pruning
+    # integrations for multi-objective optimization.
+
+    def _get_values(self) -> List[Optional[float]]:
+        trial = self._trial.study._storage.get_trial(self._trial._trial_id)
+        return [trial.intermediate_values.get(i) for i in range(self._n_objectives)]
+
+
+@experimental("1.4.0")
+class FrozenMultiObjectiveTrial(object):
+    """Status and results of a :class:`~optuna.multi_objective.trial.MultiObjectiveTrial`.
+
+    Attributes:
+        number:
+            Unique and consecutive number of
+            :class:`~optuna.multi_objective.trial.MultiObjectiveTrial` for each
+            :class:`~optuna.multi_objective.study.MultiObjectiveStudy`.
+            Note that this field uses zero-based numbering.
+        state:
+            :class:`~optuna.structs.TrialState` of the
+            :class:`~optuna.multi_objective.trial.MultiObjectiveTrial`.
+        values:
+            Objective values of the :class:`~optuna.multi_objective.trial.MultiObjectiveTrial`.
+        datetime_start:
+            Datetime where the :class:`~optuna.multi_objective.trial.MultiObjectiveTrial` started.
+        datetime_complete:
+            Datetime where the :class:`~optuna.multi_objective.trial.MultiObjectiveTrial` finished.
+        params:
+            Dictionary that contains suggested parameters.
+        distributions:
+            Dictionary that contains the distributions of :attr:`params`.
+        user_attrs:
+            Dictionary that contains the attributes of the
+            :class:`~optuna.multi_objective.trial.MultiObjectiveTrial` set with
+            :func:`optuna.multi_objective.trial.MultiObjectiveTrial.set_user_attr`.
+        intermediate_values:
+            Intermediate objective values set with
+            :func:`optuna.multi_objective.trial.MultiObjectiveTrial.report`.
+    """
+
+    def __init__(self, n_objectives: int, trial: FrozenTrial):
+        self.n_objectives = n_objectives
+        self._trial = trial
+
+        self.values = [trial.intermediate_values.get(i) for i in range(n_objectives)]
```

Comment:
It may be better to use `Tuple[float]` instead of `List[float]` as the type of `values`.

Reply:
Changed: https://github.com/optuna/optuna/pull/1054/commits/b88299a769da7b9d9ddacf0abf8af88686597b03